### ISSUE TITLE
Re-loading of an updated "Include"/partial template section?

### ISSUE THREAD

**Issue by marekvse** [toddams/RazorLight#515]
**Describe the bug**
I have a multi-level templates where the main layout has HEADE and FOOTER sections and the main layout is then used by an actual email template.

When I update the content of the template representing the FOOTER for example and even after I remove the template from the LookupCache as well as remove all the dependent templates from the LookupCache, too. Yet, opening any of the dependent layouts using the FOOTER section, they do not show the update because the callback to load the HTML for the section (from the database in my case) is never called again after it has been called just fine the first time.

I'm loading the templates from the database by their ID using the RazorLightEngineBuilder.UserProject(callbackFunction) and that works fine the first time and calls the "callbackFunction" as many times as it needs to load all the dependent layouts. But it does not call the callbackFunction again when I update an included layout and despite removing the cached items.

How do I force it to reload the source html/Razor definition again for those templates that are not in the LookupCache anymore?

**Expected behavior**
The callback function defined in RazorLightEngineBuilder.UserProject(...) should be called again if that particular (section) template is not in the Cache anymore and a layout that uses that section is loaded.

**Comment by jzabroski** [comment]
Clearing the cache and compilation are separate tasks.  It sounds like you want a CacheRefresh feature that clears the cache and re-compiles the cleared keys?

**Comment by marekvse** [comment]
That's basically right @jzabroski . Except I would not want to clear the whole compilation cache every time I update a single include. I would like to clear the cache selectively, supplying the IDs of the templates for which I want the compilation cache cleared. Then, it should add those that need to be compiled as required, as it does right now. I just want to be able to say: OK, I updated template no. 17, so clear the compilation cache for that one, but I also know that templates 23, 24, 41 and 50 directly or indirectly depend on it so clear the cache for those as well, if any of them have been compiled and stored in the compilation cache.
Clearing the compilation cache wholesale would be a "poor man's" solution as I would prefer to keep all the unaffected templates that I may want to view compiled and not waste time recompiling those, but at this point, I'd even welcome that solution. It would still be better than restarting the whole application.

**Comment by marekvse** [comment]
Yep, that's look all right. I just need something to clear the cache - per template or at worst completely.

I sure am tracking all the dependencies in the database so I know exactly what other templates I need to clear from the cache. 
I've basically build an email template editor where the user can easily switch between templates and the user may want edit the base template, like the footer, which is refrerenced by the main layout, which is then used in specific email templates. I even use more depths than that in the software I built this editor for, but that is enough to demonstrate the use case. If the user cannot see the change in the footer in this example, they will think something did not work. And the only way for them to see it is for me or the application / server admin to restart the app, which surely is not the best for usability :).

Re your EDIT: I'm not sure what you are specifically talking about there as I do not know the source code that intimately, but thre surely needs to be a way to clear the compilation cache programmatically.

**Comment by jzabroski** [comment]
I meant that the interface as a set of keys is a leaky/incorrect abstraction since the issue you are trying to address is fundamentally a dag, not a set.  Getting all snapshot keys if therefore not helpful even.  So the minimal viable product would simply be:

```c#
Task RefreshCachedTemplate(string key);
```
You would then need to sequentially call `await RefreshCachedTemplae(key)` bottom-up.
The cost of this is that others who use this will not understand that it won't work well unless they track the DAG themselves.

**Comment by marekvse** [comment]
OK, had to look up what is a DAG :). I did not go to UNI for programming so I sometimes am not aware of these generic concepts...

Anyway
```C#
Task RefreshCachedTemplate(string key);
```

sounds fine to me. Although I would prefer 
```C#
Task RemoveCachedTemplate(string key); 
```
because at the point the template (let's say the footer from my earlier example) is edited I just need to make sure that its previous compiled version is removed from the cache, not that it or any of the higher level templates that depend on it, are recompiled right then. The user may not even go back to the previously opened template that depends on it even if they did have it open before. The "Remove..." function would allow me to simply make sure that I removed all the cached templates in the tree of dependencies that have been compiled into the cache and whether or not they get re-compiled and put back into the cache later is going to be taken care of by the already existing code I believe, where if a required Include is not found in the case it is requested by the component from the code that uses it and it gets compiled then.

**Comment by marekvse** [comment]
Oh and as for the cost, I think there are lots of things that may be unclear in any component outthare. Naturally you have to read up on it or simply look into the intelisense description of the function.

I think that much more confusing, for example, is the ability to remove the templates from LookupCache which does not currently remove it from the compilation cache. I probably do not understand what the difference is there, but I for one would expect the action of removing something from LookupCache (or whatever other name it could have been called) would do what it says, which is "removing the template from the cache" and that, in my mind, should include automatically removing the template from all the caches that the component may be maintaining, so also from the compilation cache. 

So in fact if the above made sense, there would be no special need for any new function like RemoveCachedTemplate if the Remove function of the LookupCache class did this by default. In that case I would not even have to change anything except just updating the version of RazorLight. :) 
So unless there is a reason to keep the remove functions separate for those two types of caches, I'd vote for just extending the functionality of the existing remove function to make sure the cached compiled template is removed as well.

**Comment by jzabroski** [comment]
I think the takeaway is that the EngineHandler should handle template cache invalidation and forced cache eviction.  The other cache should probably only be used in the sense of relieving memory pressure from a cache eviction standpont.

Do you agree?

**Comment by marekvse** [comment]
Yes, that sounds correct to me.

Thank you for looking into this. And thank you for the two snippets above. That should help me to make the necessary changes to implement the behavior I need.

Do you think that is also something that may get implemented in the near future? My software won't be out for a while so if I knew there is a good chance this will get implemented by the original author(s) I wouldn't have to spend time fixing it for myself, where my fix may be sub-optimal.

**Comment by jzabroski** [comment]
I would love a PR.  I am pretty busy these days with a 5 month old baby boy.

**Comment by marekvse** [comment]
OK, I'll try to do that when I get around trying to fix it. However I cannot say that I'll be very confident in offering the best solution for a public component. As soon as it starts working for me I need to be on other things.
Thanks again! :)
